pstoreds/addr_book.go

```diff
@@ -2,398 +2,608 @@ package pstoreds
+		// record is not dirty, and we have no expired entries to purge.
+		return false
+	}
+
+	if len(r.Addrs) == 0 {
+		// this is a ghost record; let's signal it has to be written.
+		// Flush() will take care of doing the deletion.
+		return true
+	}
+
+	if r.dirty && len(r.Addrs) > 1 {
+		// the record has been modified, so it may need resorting.
+		// we keep addresses sorted by expiration, where 0 is the soonest expiring.
+		sort.Slice(r.Addrs, func(i, j int) bool {
+			return r.Addrs[i].Expiry < r.Addrs[j].Expiry
+		})
+	}
+
+	// since addresses are sorted by expiration, we find the first survivor and split the
+	// slice on its index.
+	pivot := -1
+	for i, addr := range r.Addrs {
+		if addr.Expiry > now {
+			break
 		}
+		pivot = i
 	}
 
-	mgr := &dsAddrBook{
-		cache:        cache,
-		ds:           store,
-		subsManager:  pstoremem.NewAddrSubManager(),
-		writeRetries: int(opts.WriteRetries),
-	}
-	return mgr, nil
+	r.Addrs = r.Addrs[pivot+1:]
+	return r.dirty || pivot >= 0
 }
 
-func keysAndAddrs(p peer.ID, addrs []ma.Multiaddr) ([]ds.Key, []ma.Multiaddr, error) {
-	var (
-		keys      = make([]ds.Key, len(addrs))
-		clean     = make([]ma.Multiaddr, len(addrs))
-		parentKey = abBase.ChildString(base32.RawStdEncoding.EncodeToString([]byte(p)))
-		i         = 0
-	)
+// dsAddrBook is an address book backed by a Datastore with a GC-like procedure
+// to purge expired entries. It uses an in-memory address stream manager.
+type dsAddrBook struct {
+	ctx context.Context
 
-	for _, addr := range addrs {
-		if addr == nil {
-			continue
-		}
+	opts Options
 
-		hash, err := mh.Sum((addr).Bytes(), mh.MURMUR3, -1)
-		if err != nil {
-			return nil, nil, err
-		}
-		keys[i] = parentKey.ChildString(base32.RawStdEncoding.EncodeToString(hash))
-		clean[i] = addr
-		i++
-	}
+	cache       cache
+	ds          ds.TxnDatastore
+	subsManager *pstoremem.AddrSubManager
 
-	return keys[:i], clean[:i], nil
-}
+	flushJobCh chan *addrsRecord
+	cancelFn   func()
+	closedCh   chan struct{}
 
-// AddAddr will add a new address if it's not already in the AddrBook.
-func (mgr *dsAddrBook) AddAddr(p peer.ID, addr ma.Multiaddr, ttl time.Duration) {
-	mgr.AddAddrs(p, []ma.Multiaddr{addr}, ttl)
+	gcCurrWindowEnd    int64
+	gcLookaheadRunning int32
 }
 
-// AddAddrs will add many new addresses if they're not already in the AddrBook.
-func (mgr *dsAddrBook) AddAddrs(p peer.ID, addrs []ma.Multiaddr, ttl time.Duration) {
-	if ttl <= 0 {
-		return
+var _ pstore.AddrBook = (*dsAddrBook)(nil)
+
+// NewAddrBook initializes a new address book given a Datastore instance, a context for managing the TTL manager,
+// and the interval at which the TTL manager should sweep the Datastore.
+func NewAddrBook(ctx context.Context, store ds.TxnDatastore, opts Options) (ab *dsAddrBook, err error) {
+	var cache cache = new(noopCache)
+	if opts.CacheSize > 0 {
+		evictCallback := func(key interface{}, value interface{}) {
+			value.(*addrsRecord).Reset()
```

Comment:
Should we be flushing dirty records here?

Reply:
This is gone now, since I removed the addrsRecord pool.